### PR TITLE
ApkGuard rule initiated

### PR DESCRIPTION
**Pull request by enovella**
New Dalvik packer added.  Still not reversed. Dumb rule at the moment. 
More details at https://github.com/rednaga/APKiD/issues/41

### CODE REVIEW

File: apkid/rules/dex/packers.yara [hunk 1]
```diff
@@ -70,3 +70,25 @@ rule medusah_appsolid_dex : packer
...
+    info = "http://apkguard.io/"
+
+  strings:
+    $method = "attachBaseContext"
+    $dexclassloader = "DexClassLoader"
+    $getclass = "getClass"
+    $getdeclaredmethod = "getDeclaredMethod"
+    $import = "java/lang/reflect/InvocationTargetException"
+    $getfilesdir = "getFilesDir"
+    $dex = "UEsDBB"
```
Comment (CalebFenton) [inline apkid/rules/dex/packers.yara]: This seems like the only thing unique to this packer. I think everything else would be really common for most packers. I'm not sure this is enough

**Comment by strazzere** [comment]
Do you have a file/hash we can look at? This mentioned issue #45 however that is for Proguard

**Comment by enovella** [comment]
It was an error linking the ticket. The one containing a sample is https://github.com/rednaga/APKiD/issues/41

File: apkid/rules/dex/packers.yara (hunk 1 above)
Comment (enovella) [reply apkid/rules/dex/packers.yara]: Hi Caleb,

I am going to try to obtain a couple of samples more. One with the same APK than before and a new one.

**Comment by strazzere** [comment]
@enovella Thanks for the commit - however, I've found some issues with it. The following hashes hit on the rule:

```
009ccf271abe87cbd0a22aa01bf277def168118a98231aa74d35378c2e3a1dec 896672b30c549c2f3a8dd22fd3722bc1b5f72dd8ba3a8757f93930d5519f9445
036392412e95dda9a7a2bd6244b2ab1b5b52f91a500d53151288d7cbd0fd1a63 8a449b0ba4b9285adfd5edcea0cb54cc6ca8c8efe427e75cc9f6bee408f12733
059e78fd265c72e5808e3244636c29fa5fd974d6cd785a58df76ca80b99af008 8a7c387e6887096761ca64e5c9b0549599cdc1f655406f0131938c1b88885b6e
07a57ae6cdd10bc4c06b8bfa016c652c0601178e8868cd368fa950ad8f7baaf4 9017f657a621e96b360abc3d43d31298b482fe255d8e6ebf4cc175ae5ab456b8
0c304874a45be19d81791a928e4bc61a1eb1860f9ec4a40737439eb9440d129b 904ee1a2dc80bd8e233a3d3605b8b73e4cadacf86794452ac359a726e03c5b98
0c49df70907c31583302727aba524466ed3bf03f57b1014c77f0736810edcbcd 905abba39a9f9e0c6a8a193ade51e74de5034a417841c99e690a536e46e25528
124d85711b268804759a26a3f993910e1763db34c0491cdc861a903b94e0629a 91b310fdbe561a74fa24c0286887fa9007e7130c6c6d019a8e13468fd25cfc37
137710d385ea86aa11675d8cdb873e0bbfc398b6507c969414eeeb71fa738605 91f1eb44a809ef84867d0dcb145b5aee8a5b62de0407927f53320211ccc71bc6
14e0b04520d0acdb35e863def4b340f54b4f4f374790038db6a0fc43fa4defb5 95f221880c42991afd9137ca135631033e02809262a5b430b1597112e856072e
1525eed1419929c6d6a7bf6a6fba13ce523ba7844879226ca3806ec2cfc3bff6 97f941acd52ad83011ccaf578059486bd8c033f9b4296ba6faa41c7b16427202
19f305072325fcd8b9a0599bbcb984234914595c35397130f3d3bfceea48ebd8 9aaf6ed90a8bb42d5feaf091385110293d0758e2391045af89d1b8f23f4beb40
1b81b41fc21d9d55af267b6b8be1bf1a5e0df40de1a3ec666a2a646d0a5b6668 9de5da14e6b85ddf05e3b0c43ce1e59db13ffcab87f66eee45695af399abfc5d
1e0fa1e00b4f2f107ce72ff299822269c5db2eb7f5a826dc494080de87a9ad21 9e85718de2004b40e2c75fc7801df7c5b9b8e5b2ad1e4dc50d5835b25911b7cc
1f40729cb3bbb57bbdfbf471acf6227544bb1b52a1dc469b721f569d42783504 9f45e794938c488b5f27bf7cdd307fab7b9126cb5d31e5319be6eefdad826b18
20023314218c29d2fbecb8b352e0c24920c7ea62bd6c9be186131eae5b467039 a06385b5af85a2aaf4fc4b2617a1be62ae12808dba1cd99bbaa2b5f258603da5
212db3269dd5e5374a07a4131f039e46fac67b0fed2e6c6862f4d4cad598ab89 a11583ca6a2963bde9f008cbad7432b11c0e8df83e81bd561c78ae19387f9368
25b8f9b4aed0d6c67ba5def2052f5d7c42ab28e078d8497c1fb4b3877f8759a7 a319725bcd4f60de8a71af1ebad0188a3d192c7bc4d922b2d68a8247b972cd73
28343baa72ba22e36fed4d7f1d0fa147e3c81a80ec947eaa219c317248b7a73a a33188ee3af71980c64aa523f589eb05409ff94e8665574e314607da118ae7c3
296cfb2d741e3fcfd05b72061e007b2e095d8e9b23eefa69a13758bfd33ac4bf a40aafad5deeeadcbf98c671b572335a5955544b2a27f463fdc66f076681a321
2c0b650e359db50a3cccc4b11073881d5bce8ce302864ce262c567702e5ecff0 a5d249a9f5f7243239a5c03507823d69242e0cc662fb6c70e42fcea7278e67ec
2f72362d1bf12e2d5b61dff53c98f422efc43000bdf31914725ea428a95830f1 b01ca9458fd4f08e6cbd258df9263819bfab73f76ac1051a3a0db14343b712c9
31a9bd84cd4d9382fdce93503c235ea16eebb27df27d2fd2ad9218c8fa8e446f b05acde57220ccf3a1e0fa833966b5a3c3ee35db9d00c7380e82f1358734d77c
31ebf19226773c5bce22f51e8a229f3a7648b5751b82865e4f2479a826d405bc b0b0997ec38bbc094243bd6bdd40e73eb506b735d2622d1e7e0739c41607173f
336ab683d287a88dd8d0051008fa7ca840b6f507bf06976b1f275078ca121b88 b2f49bf56c47cdd751740fe2a7e2a0d1566e01d13ef8a75e6fa385d91d77c63f
3402e41988f6c4fb9d8f1e05fe8965f9e9c2c1a5f6e34534e5b1336da1cc5715 b2fbd506ca0283633b41f2fa3fda355377a9901ec803e391ea7c7138f8ec010a
35cdf8232b0b7a23c30aa25eee1c1d91c582446cbb7feefe95e9590d50db3482 b3aec28e79944e295c6a4c5ae41482a4792c4a1be37e677fd25517c33ead37dc
363d341c9978410f81711258b18db853c6a89ad2d8193c4168d7b8b9f073cfba b412b21ba34c2c9eb56e6955dd7157095a5738fd67b88ecbb1229d16f5776d75
38ca538b960a29720cfaf962dc8f819c146ba5908367f11611406106d10503f8 b48ad429ab9e7402aa8da11eddb0a8251a373627a13e3ca42b743a1bbf4f33ed
3bedc9e52b21c6d5f5d14be8b30d8606b9d0b633b4bb0350066722674dc232ea b4d099e67909df49eaf415e76e505424a91c703bcec31f9f0494c7de0e8d83ce
3eec4d96f2240719545452a4fb79ca576cca5f6f711c5e9bdf7c6954e6a301de b5611bbae4f0896ef721f6ca07db2447d43bbd8344d578f910308d0cc4ea0dac
3fec8f0ffc1c4f4be39d970e1bfc19630c478b3945749d23272c3d9da40ba6b7 b5820044f73e3f4b076b6518a5202b78e60dc010250d9134df77c0b403ee0071
424ee1378970b745e46dd9df116f0598aea99aba098fcdf3ae4086e11aca8a99 b76f50c115ca27e29959af73dae67cd092586684765fb1e8ab1818df11693a5d
43603f58f8f7e317c7244172c312408cef8487151509ba4e074b38db5894b31d b7ce3f273a1d4d5c9dec63df9360aefe52c443c070339f2d1cab4592a6f023f6
470ff3775eebcef684615b07e7413276cfbf4a0cfa6baf60898dcedfcd062a13 b8d1e3d6831a12d20995ebabfecaf4f0c3a816d920aa3fb388469763a285db9c
4b81d30f81655863f19cc606b4493716b69948c80dbe6a4c333b6a9451ed957d ba60c0ef259bd28f4288bf5be9fd0aeecd44629193685fb8e1aecbe6815591c7
4c14e5043402b21a6fd44a5539a266767b42fc73b177537527662a03915d8a69 bc4fdc89f586ab1b0c42b3221f6ea0ab89543014dc95f1c534a0ed5721f3e50d
4d37f32476859d233c0649763fb3d602abc01dc071ea049391aa91cfc46e134a bd6d4291f32fb679593097420fe0f90f703f1cf0fcaa35b9f574812062e61bab
4f2809e8fa5b5756590fb6375c2522109f9ce4f581ff1fa31256cafb456f8876 bdef1316a06d6ee8f18c11b35b049c0870349d3d9dea84e0ba507fcacef07538
5126ebede7e34a6e18474fd47681798e648059c43c376cfd9ab311d20d2416b7 be151ba2a1a8d81f0f2ad6533977dce0a52f102bb6dd4c1da5749bd801b31600
51843f098c791337eaeece35696f7208aec84888807ed415012c22f4dee2fa4b c05e0eb6820c12f79fa9b2e4a980f1607e19b4fe3d880d766825181a7ccb6450
51a807dbfdc7045146ce001794309d3f81db307bbffc3947340c1cfdac1f2a80 c41420067f9574d3026766aa27aec4aa3eef0688fb7c6b5664decf80b9b9d840
5310ac5d239953fbd0b2d739964a0a8b56ae45e6503cc3260095213a8e1dac07 c50905330fe7735535e1ca1c227baff50512875984d3eb07a74eab7be4f34012
5431b596ead2bec5291eea78621db8fc101ae8cf1f948e1085f4f7ed992d45bc c621ab6794977ed732c71260a74536d2f02d9fdc9f4bbd6c66dad4d94f8a2ba6
546972a8ca05e77669ab552ab9d3e624f899eac7566aef4607e0f1f899c76e02 c6557715a12a48e454d3e96395045a42ed415101a360ce68ec757286fe960465
54b07f2d48cfe12eb6c895c3376f8803723fbcfc6c3c2af63370bb1061d67f6b c71c25d3b95c2a97a4fa3585c9025711b24bffea741a450a9486833b4a66fba1
5950348a66e61ed968cf50fc8c5546f6de65f0e32cf2f32df8bca3d8874a14b9 c85182589dd60263afaf85b9e931b767e15c5f2370ffc824c1db521957a50f66
5da0c345cd20a2c255f6e5596551f2e83a48ddeda3956928d6498c6287e2dda5 ca8b2becfebdf5b47f0416eff8c2f21e4818d8d2d3a3ea786232bb1cd9cd16fd
5e95b1d2b21a7975bd48b50e39b9fa29942f0ab6cb123c2a07cf2899c4a4e94b ccd35d8e700d6bb9d4096fad0020dd025edd5f3f9cfc04629c4220e8721444be
6233cf54a3808556d544bf2bdd1133f039659a74d4907008762b35d07f290a59 cdf5eaa8aa97b3b870c4bb605c27fc9c684a9521aaf7c9767763415e3b92ba63
63717a5437e03a24ee96b38706dbc5bca6047b80729415ff3e4a26f57f5e0e5d d4c722b555c3020ddb146ac8cdaf2549d05f4ba1f87d0a59b132f9130970f6d4
6578b67cf6dd5539eca3267ede730fa88624243c3a6d803f94f530cd1ed6bd3d d5f0edabb0c9a383f85a4ad00b170a9188d356e4286cc53c07c8e6925c04198d
659c27d6170b8c0686727144ffa89275d4485f992a064e2d28df337d0c3f9f1f d64f708ac54bf8d4a48a89b2e55b8fbd4e94efeae2949007fe11dc15ff71d942
65d26941b4ff55b64bd30b41a79e550d2a9de8aae9decaa6c848cf568587baee dea6fe02c662f0a12f9292ea0bafda6d4bcf762a56ed7b88de90abef068686ad
6824084975fbfbc18ac9ca05ba6b7932c94fa03502dda5d8ae992344dae3a6f0 debf13c65f3d821eec6e40a35645e28d37fec790a46c21fdbc74bbc898c2cc29
6943bce2189650eb685bf7d367c0bb8a60481ea2be3dae14f0304a6205b33968 e19dea08f6658afe0bf16773cc1dd64ba9360439c417f2dd25cf8e0aa1854b96
6c085ac1bbbfa9546e996898b5658a31a8079e31144e4c00242a1e0cc9fea617 e1e5d3fdbcdde742451f84b9bd2cb1b5951114536a931ad5166f6f7c32566d02
6e945e99ab59e6ab8d62ef0a5b410b08f0b434481df6e21e9f558cd055eff169 e8c95f3da4bbb9371c9f05a86e50465b4373f46a14e0d2057ecde38af3c66048
719a1512978a81f1f049e5f06986e64425f75528c926bbb0b4170f29d539dab9 e97ed6a77b420be651087e40210721c7a5aac8ccacef1e7ef50fff65d6083b1c
72ab61babdae6d0391996e4416adfa7d8270ded167ad8b0a3d851a502400b5d8 e9c68ad5b6157889c90ba379f5cabb5bf2d5caf050a7c5401e8a24932bfbcfbd
7319fab8998e2f41494382513fc3ee73b06127471242b8d8ce3b7e258111cfaf ebf4e310aa1d64cb57f4efa4e165108e8f9ca4980e9978edf2d596e76f5d9468
782e0af589f9bfe7a302860ef141c595f0b24a3b7d26278bf4803083ecf8efd8 f174de80997cdc4a1bd452f02d36a18f5232df8ec3363cda28632fece831ee2c
7daccb2bafb8ba66ec3936aa81252d2c7887af17cff43ca92a943d4507b52104 f210c96a5a2c258d49580920bf2fd4cfc050961c2c8f25b9dbad0985ba3d5ad0
7e6bb75074c7e31e5abed3ea090f982a323f589e491d0ca0e22bb8c7842cd3b3 f340ed5023eb56782b2ddb0813e71d3017a21326379f79b2f986d6660ab95d52
7fffabebb049fddf749a7818863ee78defc9fbfee88faedd801ff7a2a64657d1 f36faa2ee08ed4634733d1ee2535f72698b11805b5c3ae5afdd51a177592528f
81a6d84747c5719bcd7203d1fb5bb649f3da53b7cb06aa8bce61ac334741447c f749f4fd5ffceca58e6d38d1d65598d1d04cdf49dc44f8344264e33c2afbc17e
82018d4231c1b5995451a13333de060e14182f5ae2cd794e818c3d2fba9410f0 f924194d5afc2c0a00f4542ec706418c6d7efc9c8f835d391f0245d7b62a522d
853683f7b817c0f57b43009361cbcf12f71dccd973c2d970b362d2924050dfa9 fd4b7b9d178ee6c89394e12310c118b4551361c0d6704d6c56dd72aef824f56b
```

However, they are not APKGuarded it would seem.

It did find the following hashes which we didn't previously know about;
```
42040d2f9c30371c4019746b70e465bac92b32c27be2448a2259e501ce3f5826
d2a6cbe9acd4193188f7aa6d922c916999845da82171889526550790f5632b47
d6e48539252c4425bbb8f4b7e60f9ca6cbb703f324bbf1dde025a3d935b74cb9
486447ed5abcf3c2c254330bc5d5a1bfd1d3b22d72d54060bb5512c0fe352944
faf4ebf16cf88bffec327cd53694421659ac3b8101ef525a9ac87236aa20cd6b
cd4ed2325daaa9dc65744fceff77be32b4166cb3e0021c278a9319048cb112d3
```

I have created a rule that would hit these. The concept I took to create these was performing a `dexdump -d` on the two samples you provided and took the opcodes less the types/methods/strings. This resulted in a pretty good set of bytes which we can match consistently:

```
        120b            // const/4 v11, #int 0 // #0
        6f20 0100 fe00  // invoke-super {v14, v15}, Landroid/app/Application;.attachBaseContext:(Landroid/content/Context;)V // method@0001
        2206 ??00       // new-instance v6, Ljava/io/File; // type@0006
        6e10 ??00 0e00  // invoke-virtual {v14}, Lyxlhycuqv/weudayy;.getFilesDir:()Ljava/io/File; // method@0019
        0c09            // move-result-object v9
        1a0a 2f00       // const-string v10, "lllllllllllllllllllllllllllllllllllllllll.zip" // string@002f
        7030 ??00 960a  // invoke-direct {v6, v9, v10}, Ljava/io/File;.<init>:(Ljava/io/File;Ljava/lang/String;)V // method@000a
        1a09 1900       //  const-string v9, BASE64_ENCODED_ZIP_FILE
        7120 ??00 b900  // invoke-static {v9, v11}, Landroid/util/Base64;.decode:(Ljava/lang/String;I)[B // method@0003
        0c02            // move-result-object v2
        2205 ??00       // new-instance v5, Ljava/io/FileOutputStream; // type@0007
        7020 ??00 6500  // invoke-direct {v5, v6}, Ljava/io/FileOutputStream;.<init>:(Ljava/io/File;)V // method@000c
        2201 ??00       // new-instance v1, Ljava/io/BufferedOutputStream; // type@0005
        7020 ??00 5100  // invoke-direct {v1, v5}, Ljava/io/BufferedOutputStream;.<init>:(Ljava/io/OutputStream;)V // method@0006
        6e20 ??00 2100  // invoke-virtual {v1, v2}, Ljava/io/BufferedOutputStream;.write:([B)V // method@0009
        6e10 ??00 0100  // invoke-virtual {v1}, Ljava/io/BufferedOutputStream;.flush:()V // method@0008
        6e10 ??00 0100  // invoke-virtual {v1}, Ljava/io/BufferedOutputStream;.close:()V // method@0007
        6e10 ??00 0600  // invoke-virtual {v6}, Ljava/io/File;.getAbsolutePath:()Ljava/lang/String; // method@000b
        0c03            // move-result-object v3
        6e10 ??00 0e00  // invoke-virtual {v14}, Lyxlhycuqv/weudayy;.getFilesDir:()Ljava/io/File; // method@0019
        0c09            // move-result-object v9
        6e10 ??00 0900  // invoke-virtual {v9}, Ljava/io/File;.getAbsolutePath:()Ljava/lang/String; // method@000b
        0c07            // move-result-object v7
        6e10 ??00 0e00  // invoke-virtual {v14}, Lyxlhycuqv/weudayy;.getClassLoader:()Ljava/lang/ClassLoader; // method@0018
        0c00            // move-result-object v0
        2204 ??00       //  new-instance v4, Ldalvik/system/DexClassLoader; // type@0004
        1209            // const/4 v9, #int 0 // #0
        7050 ??00 3497  // invoke-direct {v4, v3, v7, v9, v0}, Ldalvik/system/DexClassLoader;.<init>:(Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;Ljava/lang/ClassLoader;)V // method@0004
        1a09 ??00       // const-string v9, "yabno/blkngwigpd" // string@003d
        6e20 ??00 9400  // invoke-virtual {v4, v9}, Ldalvik/system/DexClassLoader;.loadClass:(Ljava/lang/String;)Ljava/lang/Class; // method@0005
        0c09            // move-result-object v9
        120a            // const/4 v10, #int 0 // #0
        23aa ??00       // new-array v10, v10, [Ljava/lang/Class; // type@0016
        6e20 ??00 a900  // invoke-virtual {v9, v10}, Ljava/lang/Class;.getConstructor:([Ljava/lang/Class;)Ljava/lang/reflect/Constructor; // method@000d
        0c09            // move-result-object v9
        120a            // const/4 v10, #int 0 // #0
        23aa ??00       // new-array v10, v10, [Ljava/lang/Object; // type@0017
        6e20 ??00 a900  // invoke-virtual {v9, v10}, Ljava/lang/reflect/Constructor;.newInstance:([Ljava/lang/Object;)Ljava/lang/Object; // method@0013
        0c09            // move-result-object v9
        5be9 0000       // iput-object v9, v14, Lyxlhycuqv/weudayy;.aaa:Ljava/lang/Object; // field@0000
        54e9 0000       // iget-object v9, v14, Lyxlhycuqv/weudayy;.aaa:Ljava/lang/Object; // field@0000
        6e10 ??00 0900  // invoke-virtual {v9}, Ljava/lang/Object;.getClass:()Ljava/lang/Class; // method@0012
        0c09            // move-result-object v9
        1a0a ??00       // const-string v10, "attachBaseContext" // string@0022
        121b            // const/4 v11, #int 1 // #1
        23bb ??00       // new-array v11, v11, [Ljava/lang/Class; // type@0016
        120c            // const/4 v12, #int 0 // #0
        1c0d ??00       // const-class v13, Landroid/content/Context; // type@0002
        4d0d 0b0c       // aput-object v13, v11, v12
        6e30 ??00 a90b  // invoke-virtual {v9, v10, v11}, Ljava/lang/Class;.getDeclaredMethod:(Ljava/lang/String;[Ljava/lang/Class;)Ljava/lang/reflect/Method; // method@000e
        0c09            // move-result-object v9
        54ea 0000       // iget-object v10, v14, Lyxlhycuqv/weudayy;.aaa:Ljava/lang/Object; // field@0000
        121b            // const/4 v11, #int 1 // #1
        23bb ??00       // new-array v11, v11, [Ljava/lang/Object; // type@0017
        120c            // const/4 v12, #int 0 // #0
        4d0e 0b0c       // aput-object v14, v11, v12
        6e30 ??00 a90b  // invoke-virtual {v9, v10, v11}, Ljava/lang/reflect/Method;.invoke:(Ljava/lang/Object;[Ljava/lang/Object;)Ljava/lang/Object; // method@0015
        0e00            // return-void
        0d08            // move-exception v8
        6e10 ??00 0800  // invoke-virtual {v8}, Ljava/lang/Exception;.printStackTrace:()V // method@000f
        28fb            // goto 0073 // -0005
```

I'll commit this and close the other issues.

**Comment by enovella** [comment]
Bad news!

I have found a new sample that doesn't match the opcode-like rule but it does on the initial & dumb rule. As discussed over slack, will share the sample privately with you guys.

**Comment by enovella** [comment]
> It did find the following hashes which we didn't previously know about;
> 
> 42040d2f9c30371c4019746b70e465bac92b32c27be2448a2259e501ce3f5826
> d2a6cbe9acd4193188f7aa6d922c916999845da82171889526550790f5632b47
> d6e48539252c4425bbb8f4b7e60f9ca6cbb703f324bbf1dde025a3d935b74cb9
> 486447ed5abcf3c2c254330bc5d5a1bfd1d3b22d72d54060bb5512c0fe352944
> faf4ebf16cf88bffec327cd53694421659ac3b8101ef525a9ac87236aa20cd6b
> cd4ed2325daaa9dc65744fceff77be32b4166cb3e0021c278a9319048cb112d3
> 

OMG, I didn't read this message from you. They seem to be packed in the same manner, right?  Will review it during the weekend and will try to adjust your opcode-like rule to hit the new sample too.

**Pull request closed without merging**